platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java

```diff
@@ -80,7 +82,16 @@ private ExecutionOutputState captureOutputsAfterExecution(UnitOfWork work, C con
                 // As this is _roughly_ the amount of time that is avoided by reusing the outputs,
                 // which is currently the _only_ thing this value is used for.
                 Duration originExecutionTime = result.getDuration().plus(Duration.ofMillis(snapshotOutputDuration));
-                OriginMetadata originMetadata = new OriginMetadata(buildInvocationScopeId.asString(), originExecutionTime);
+                HashCode buildCacheKey = context.getCachingState()
+                    .fold(
+                        (CachingState.Enabled enabled) -> Optional.of(enabled.getKey()),
+                        CachingState.Disabled::getKey)
+                    .map(cacheKey -> HashCode.fromBytes(cacheKey.toByteArray()))
+                    .orElse(null);
```

Comment:
:x: We should never get `null` here. The problem is that the cache key does not belong to the caching state anymore after your changes. It should instead be moved to the `BeforeExecutionState`, and should be calculated there. Once we do that, it will always be available when a `BeforeExecutionState` is captured.

That would also allow `OriginMetadata.getHashCode()` to be non-null, as it should be.

Reply:
There's one more problem with doing this, which is `SkipEmptyNonIncrementalWorkStep`. I think in the long term we should stop supporting it (maybe in 9.0). Until then we can move it after `CaptureNonIncrementalStateBeforeExecutionStep`, and thus get a build cache key captured.

Reply:
We don't always calculate the cache key when e.g. build caching is off. Though we always create the origin metadata.
You can see this from the tests as well: When build caching is off and the scan plugin isn't applied, the cache key will be null here.

Reply:
Hmm... It seems that we already don't support skip when empty for non-incremental work: https://github.com/gradle/gradle/blob/c399528b58d980780c234330436d4890bd170688/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DependencyManagementGlobalScopeServices.java#L123-L173

Am I missing something?

Reply:
Alright, now the `cacheKey` is non-null in `OriginMetadata` 🎉

Reply:
Yay!